ravendb/rvn_server.go

```diff
@@ -0,0 +1,824 @@
+package ravendb
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/ravendb/ravendb-go-client"
+	"github.com/ravendb/ravendb-go-client/serverwide/operations"
+	"github.com/ravendb/terraform-provider-ravendb/utils"
+	"golang.org/x/crypto/ssh"
+	"log"
+	"net"
+	"net/url"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	numberOfIterationsToTryAndGetTopology int = 5
+	numberOfIterationsExecuteWithRetries  int = 5
+	DEFAULT_SECURE_RAVENDB_HTTP_PORT      int = 443
+	DEFAULT_INSECURE_RAVENDB_HTTP_PORT    int = 8080
+	DEFAULT_SECURE_RAVENDB_TCP_PORT       int = 38888
+	DEFAULT_INSECURE_RAVENDB_TCP_PORT     int = 38881
+	DEFAULT_HTTP_PORT                     int = 80
+)
+
+type ServerConfig struct {
+	Package             Package
+	Hosts               []string
+	License             []byte
+	Settings            map[string]interface{}
+	ClusterCertificate  []byte
+	Url                 Url
+	Assets              map[string][]byte
+	Insecure            bool
+	SSH                 SSH
+	HealthcheckDatabase string
+}
+
+type NodeState struct {
+	Host               string
+	Licence            []byte
+	Settings           map[string]interface{}
+	ClusterCertificate []byte
+	HttpUrl            string
+	TcpUrl             string
+	Assets             map[string][]byte
+	Insecure           bool
+	Version            string
+	Failed             bool
+}
+
+type Package struct {
+	Version string
+	Arch    string
+}
+
+type Url struct {
+	List     []string
+	HttpPort int
+	TcpPort  int
+}
+
+type SSH struct {
+	User string
+	Pem  []byte
+	Port int
+}
+
+func (s *SSH) getPort() int {
+	if s.Port != 0 {
+		return s.Port
+	}
+	return 22
+}
+
+type DeployError struct {
+	Output string
+	Err    error
+}
+
+func (e *DeployError) Error() string {
+	return e.Err.Error() + " with output:\n" + e.Output
+}
+
+func upload(con *ssh.Client, buf bytes.Buffer, path string, content []byte) error {
```

Comment:
Maybe add a user that we'll set as the owner of the file after the upload? Given that we `sudo scp` ?